Userland/Libraries/LibMarkdown/Text.cpp

```diff
@@ -8,268 +9,357 @@
+    bool is_image = opening == "![";
 
-    for (size_t offset = 0; offset < str.length(); offset++) {
-        char ch = str[offset];
+    auto link_text = parse_sequence(tokens, true);
 
-        bool is_escape = ch == '\\';
-        if (is_escape && offset != str.length() - 1) {
-            offset++;
-            continue;
-        }
+    if (tokens.past_end() || *tokens != "](") {
+        link_text->children.prepend(make<TextNode>(opening.data));
+        return link_text;
+    }
+    auto seperator = *tokens;
+    VERIFY(seperator == "](");
 
-        bool is_special_character = false;
-        is_special_character |= ch == '`';
-        if (!current_style.code)
-            is_special_character |= ch == '*' || ch == '_' || ch == '[' || ch == ']' || (ch == '!' && offset + 1 < str.length() && str[offset + 1] == '[');
-        if (!is_special_character)
-            continue;
-
-        append_span_if_needed(offset);
-
-        switch (ch) {
-        case '`':
-            current_style.code = !current_style.code;
-            break;
-        case '*':
-        case '_':
-            if (offset + 1 < str.length() && str[offset + 1] == ch) {
-                offset++;
-                current_style.strong = !current_style.strong;
-            } else {
-                current_style.emph = !current_style.emph;
-            }
-            break;
-        case '!':
-            current_link_is_actually_img = true;
-            break;
-        case '[':
-            if constexpr (MARKDOWN_DEBUG) {
-                if (first_span_in_the_current_link != -1)
-                    dbgln("Dropping the outer link");
-            }
-            first_span_in_the_current_link = spans.size();
-            break;
-        case ']': {
-            if (first_span_in_the_current_link == -1) {
-                dbgln_if(MARKDOWN_DEBUG, "Unmatched ]");
-                continue;
-            }
-            ScopeGuard guard = [&] {
-                first_span_in_the_current_link = -1;
-                current_link_is_actually_img = false;
-            };
-            if (offset + 2 >= str.length() || str[offset + 1] != '(')
-                continue;
-            offset += 2;
-            size_t start_of_href = offset;
-
-            do
-                offset++;
-            while (offset < str.length() && str[offset] != ')');
-            if (offset == str.length())
-                offset--;
-
-            const StringView href = str.substring_view(start_of_href, offset - start_of_href);
-            for (size_t i = first_span_in_the_current_link; i < spans.size(); i++) {
-                if (current_link_is_actually_img)
-                    spans[i].style.img = href;
-                else
-                    spans[i].style.href = href;
-            }
-            break;
-        }
-        default:
-            VERIFY_NOT_REACHED();
+    auto address = make<MultiNode>();
+    for (auto iterator = tokens + 1; !iterator.past_end(); ++iterator) {
```

Comment:
`tokens` was just dereferenced, therefore it is not `.is_end()`. Therefore, `tokens + 1` is either `.is_end()` or earlier.

Why is `past_end()` necessary here? And in case my logic is wrong: Either fix the previous access, or wrap the six-line for-loop in an `if`, so that we can avoid beyond-the-end-iterators.

Reply:
`past_end()` has been removed